### ISSUE TITLE
Bogart ignoring memory allocation

### ISSUE THREAD

**Issue by cmfield** [marbl/canu#1253]
I am submitting a job to our SGE that runs canu with useGrid=False, and no matter what memory I throw at it, or what setting I try, I get the same error, that there is not enough memory. It claims it is limited to 16G (user supplied) but I have never specified that amount. It is a metagenomic assembly. Here is my run command:

canu -d canu_asm -p asm_180219 genomeSize=5m -useGrid=false -maxThreads=24 -maxMemory=768 -corOutCoverage=10000 -corMhapSensitivity=high -corMinCoverage=0 -pacbio-raw <files>

**And the log:**

```
-- Local: cormhap    6 GB   12 CPUs x   2 jobs    12 GB   24 CPUs  (overlap detection with mhap)
-- Local: obtovl     4 GB    8 CPUs x   3 jobs    12 GB   24 CPUs  (overlap detection)
-- Local: utgovl     4 GB    8 CPUs x   3 jobs    12 GB   24 CPUs  (overlap detection)
-- Local: ovb        4 GB    1 CPU  x  24 jobs    96 GB   24 CPUs  (overlap store bucketizer)
-- Local: ovs        8 GB    1 CPU  x  24 jobs   192 GB   24 CPUs  (overlap store sorting)
-- Local: red        8 GB    4 CPUs x   6 jobs    48 GB   24 CPUs  (read error detection)
-- Local: oea        4 GB    1 CPU  x  24 jobs    96 GB   24 CPUs  (overlap error adjustment)
-- Local: bat       16 GB    4 CPUs x   1 job     16 GB    4 CPUs  (contig construction with bogart)
-- Local: gfa        8 GB    4 CPUs x   1 job      8 GB    4 CPUs  (GFA alignment and processing)
--
-- In 'asm_180219.seqStore', found PacBio reads:
--   Raw:        1837959
--   Corrected:  1836747
--   Trimmed:    904350
--
-- Generating assembly 'asm_180219' in '/nfs/cds-shini.ethz.ch/exports/biol_micro_cds_gr_sunagawa/scratch/chris/jcahn/canu_asm'
--
-- Parameters:
--
--  genomeSize        5000000
--
--  Overlap Generation Limits:
--    corOvlErrorRate 0.2400 ( 24.00%)
--    obtOvlErrorRate 0.0450 (  4.50%)
--    utgOvlErrorRate 0.0450 (  4.50%)
--
--  Overlap Processing Limits:
--    corErrorRate    0.3000 ( 30.00%)
--    obtErrorRate    0.0450 (  4.50%)
--    utgErrorRate    0.0450 (  4.50%)
--    cnsErrorRate    0.0750 (  7.50%)
--
--
-- BEGIN ASSEMBLY
--
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'bat' concurrent execution on Wed Feb 20 09:11:06 2019 with 166849.542 GB free disk space (1 processes; 1 concurrently)

    cd unitigging/4-unitigger
    ./unitigger.sh 1 > ./unitigger.000001.out 2>&1

-- Finished on Wed Feb 20 09:11:06 2019 (furiously fast) with 166849.542 GB free disk space
----------------------------------------
--
-- Bogart failed, retry
--
--
-- Running jobs.  Second attempt out of 2.
----------------------------------------
-- Starting 'bat' concurrent execution on Wed Feb 20 09:11:06 2019 with 166849.542 GB free disk space (1 processes; 1 concurrently)

    cd unitigging/4-unitigger
    ./unitigger.sh 1 > ./unitigger.000001.out 2>&1

-- Finished on Wed Feb 20 09:11:07 2019 (one second) with 166849.542 GB free disk space
----------------------------------------
--
-- Bogart failed, tried 2 times, giving up.
```

**And finally the unitigger log:**

```
==> PARAMETERS.

Resources:
  Memory                16 GB
  Compute Threads       4 (command line)

Lengths:
  Minimum read          0 bases
  Minimum overlap       500 bases

Overlap Error Rates:
  Graph                 0.045 (4.500%)
  Max                   0.045 (4.500%)

Deviations:
  Graph                 6.000
  Bubble                6.000
  Repeat                3.000

Edge Confusion:
  Absolute              2100
  Percent               200.0000

Unitig Construction:
  Minimum intersection  500 bases
  Maxiumum placements   2 positions

Debugging Enabled:
  (none)

==> LOADING AND FILTERING OVERLAPS.

ReadInfo()-- Using 1837959 reads, no minimum read length used.

OverlapCache()-- limited to 16384MB memory (user supplied).

OverlapCache()--      14MB for read data.
OverlapCache()--      70MB for best edges.
OverlapCache()--     182MB for tigs.
OverlapCache()--      49MB for tigs - read layouts.
OverlapCache()--      70MB for tigs - error profiles.
OverlapCache()--    4096MB for tigs - error profile overlaps.
OverlapCache()--       0MB for other processes.
OverlapCache()-- ---------
OverlapCache()--    4516MB for data structures (sum of above).
OverlapCache()-- ---------
OverlapCache()--      35MB for overlap store structure.
OverlapCache()--   11832MB for overlap data.
OverlapCache()-- ---------
OverlapCache()--   16384MB allowed.
OverlapCache()--
OverlapCache()-- Retain at least 1133 overlaps/read, based on 566.50x coverage.
OverlapCache()-- Initial guess at 421 overlaps/read.
OverlapCache()--
OverlapCache()-- Not enough memory to load the minimum number of overlaps; increase -M.
```

**Comment by skoren** [comment]
The default limits are set based on the genome size. For metagenomic assembly this is often not sufficient so you need explicitly increase it (see the FAQ: https://canu.readthedocs.io/en/latest/faq.html#what-parameters-can-i-tweak). Adding ```batMemory=200```should let you run.

**Comment by cmfield** [comment]
It doesn't matter what I set -batMemory to, I get the exact same error log. It still says it's limited to 16G. I even tried with just -batMemory=200 and no -maxMemory in case they were clashing somehow.

**Comment by skoren** [comment]
Remove the 4-unitigger folder first, then re-launch with batMemory. Post the log from that output, especially the preamble where canu lists the resources it will use.

**Comment by cmfield** [comment]
That worked. There must have been something there using previous settings. The first time I ran I had -maxMemory=96 -maxThreads=32, so was the 16G limit from those settings or should I always specify -batMemory separately in future?

Log for reference:

```
-- Detected Java(TM) Runtime Environment '1.8.0_192' (from '/nfs/nas22/fs2201/biol_micro_unix_modules/modules/software/Java/1.8.0_192/bin/java') with -d64 support.
-- Detected gnuplot version '5.2 patchlevel 5   ' (from 'gnuplot') and image format 'png'.
-- Detected 144 CPUs and 1007 gigabytes of memory.
-- Limited to 768 gigabytes from maxMemory option.
-- Limited to 24 CPUs from maxThreads option.
-- Detected Sun Grid Engine in '/opt/sge/default'.
-- Grid engine disabled per useGrid=false option.
--
--                            (tag)Concurrency
--                     (tag)Threads          |
--            (tag)Memory         |          |
--        (tag)         |         |          |     total usage     algorithm
--        -------  ------  --------   --------  -----------------  -----------------------------
-- Local: meryl     12 GB    4 CPUs x   6 jobs    72 GB   24 CPUs  (k-mer counting)
-- Local: hap        8 GB    4 CPUs x   6 jobs    48 GB   24 CPUs  (read-to-haplotype assignment)
-- Local: cormhap    6 GB   12 CPUs x   2 jobs    12 GB   24 CPUs  (overlap detection with mhap)
-- Local: obtovl     4 GB    8 CPUs x   3 jobs    12 GB   24 CPUs  (overlap detection)
-- Local: utgovl     4 GB    8 CPUs x   3 jobs    12 GB   24 CPUs  (overlap detection)
-- Local: ovb        4 GB    1 CPU  x  24 jobs    96 GB   24 CPUs  (overlap store bucketizer)
-- Local: ovs        8 GB    1 CPU  x  24 jobs   192 GB   24 CPUs  (overlap store sorting)
-- Local: red        8 GB    4 CPUs x   6 jobs    48 GB   24 CPUs  (read error detection)
-- Local: oea        4 GB    1 CPU  x  24 jobs    96 GB   24 CPUs  (overlap error adjustment)
-- Local: bat      200 GB    4 CPUs x   1 job    200 GB    4 CPUs  (contig construction with bogart)
-- Local: gfa        8 GB    4 CPUs x   1 job      8 GB    4 CPUs  (GFA alignment and processing)
--
-- In 'asm_180219.seqStore', found PacBio reads:
--   Raw:        1837959
--   Corrected:  1836747
--   Trimmed:    904350
--
-- Generating assembly 'asm_180219' in '/nfs/cds-shini.ethz.ch/exports/biol_micro_cds_gr_sunagawa/scratch/chris/jcahn/canu_asm'
--
-- Parameters:
--
--  genomeSize        5000000
--
--  Overlap Generation Limits:
--    corOvlErrorRate 0.2400 ( 24.00%)
--    obtOvlErrorRate 0.0450 (  4.50%)
--    utgOvlErrorRate 0.0450 (  4.50%)
--
--  Overlap Processing Limits:
--    corErrorRate    0.3000 ( 30.00%)
--    obtErrorRate    0.0450 (  4.50%)
--    utgErrorRate    0.0450 (  4.50%)
--    cnsErrorRate    0.0750 (  7.50%)
--
--
-- BEGIN ASSEMBLY
--
-- Finished stage 'unitig', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'bat' concurrent execution on Wed Feb 20 10:15:21 2019 with 166715.314 GB free disk space (1 processes; 1 concurrently)

    cd unitigging/4-unitigger
    ./unitigger.sh 1 > ./unitigger.000001.out 2>&1
```

**Comment by skoren** [comment]
No it's set from the genome size which normally is sufficient, you should specify it directly for metagenomic datasets since the genome size isn't actually reflective of the expected number of overlaps as documented on the FAQ.
